ROADMAP.md

```diff
@@ -0,0 +1,579 @@
+
+## Visionary
+
+**Our long term roadmap**.
+
+This is the stuff that moves libp2p from "a networking toolbox to build P2P
+applications" to the thing that fundamentally reshapes the architecture of the
+Internet; our dreams and aspirations, the North star we should always keep in
+sight; this is what motivates us and it's speaks intimately to our mission
+statement; the libp2p analogy of IPFS working on Mars
+
+### 🖧 Decentralizing networks
+
+**What?** Interacting and transacting with the *person next to me* should not
+require me to resort to the Internet, make use of centralised services run by
+corporations, expose myself to censorship, and lose self-agency over my data.
+
+This theme covers a number of related topics: mesh networking, community
+networks, sneaker nets, isolated networks in libp2p.
+
+An undertaking like this touches upon many surfaces of the system, e.g.
+implementing new transports (Bluetooth, software defined radio,
+ultrasonic data \[Chirp\], Wi-Fi direct, QR, Ethernet, 802.11s, etc.),
+new discovery mechanisms (local peer exchange, social peer
+advertisement, etc.), packet-switched routing, and more.
+
+But it's not just about connectivity. In this new reality, nodes can
+come and go easily. As a result, new requirements will emerge, such as
+supporting *roaming* (the ability for a user to switch networks without
+dropping ongoing sessions) and *resilient state and connectivity*, to
+buffer data (potentially via a store-and-forward architecture) so that a
+node disconnecting intermittently can get up to speed when they come
+back online.
+
+For certain applications, we might want to take compensatory actions
+after a grace period, such as rebalancing data across nodes. And this
+requires us to think about group membership, partitioning, consensus,
+etc. This is just an initial train of thought, and there's a lot more
+ground to cover.
+
+**Why?** The IPFS roadmap tackles mesh networks, isolated networks,
+sneaker nets, and making IPFS work in Mars where general
+connectedness to the Internet does not exist. *The network should pave
+the way for IPFS to evolve in that direction.*
+
+### 🤫 A spyproof libp2p
+
+**What?** Supporting transports like cjdns, I2P and Tor (see
+[OpenBazaar/go-onion-transport](https://github.com/OpenBazaar/go-onion-transport))
+helps create resilient, privacy-centric communications, and moves the needle
+forward towards censorship-resistant networks. As an intermediary step, we
+should improve our existing transport stack, e.g. preventing downgrade attacks
+when securing connections.
+
+### 📱 libp2p in mobile devices
+
+**What?** While js-libp2p technically works in mobile browsers, it is
+constrained by the web sandbox. First-class integration of native libp2p
+in mobile devices could be a prerequisite for the class of offline-first
+use cases we are targeting at Protocol Labs.
+
+This implies building/supporting Java/Kotlin and Swift implementations
```

Comment:
Are any customers asking for this?  Have any contributors created these?  If not, how come?

Reply:
I am not particularly familiar with this part of the libp2p space. I know there is a [JVM libp2p implementation](https://github.com/libp2p/jvm-libp2p).

I don't think we, as in Protocol Labs have the bandwidth to maintain either of these. That said I do think we can support other teams building these, either from scratch or wrapping existing implementations like rust-libp2p (added via 5aed005cf4085b0c5a86af843d3533a682cd0885).

Reply:
Native support for a lot of android devices at least is possible with https://github.com/status-im/nim-libp2p - I imagine that can be extended to OSX as well

Reply:
rust-libp2p already works on android and iOS and has been for years. Both sunshine and actyx actually build Android apps and sunshine had an iOS app.

Reply:
Not that it is a huge accomplishment, rust literally just runs everywhere. This also doesn't mean that there isn't room for improvement. Adding ci to rust-libp2p for example to prevent regressions like: https://github.com/libp2p/rust-libp2p/issues/2064

Reply:
At the risk of creating a language shootout, the main point of my comment was that a Java/Kotlin fork is not a prerequisite ;)

Reply:
> Adding ci to rust-libp2p for example to prevent regressions like: libp2p/rust-libp2p#2064

As an aside, feel free to add a CI step to rust-libp2p to prevent regressions going forward @dvc94ch and @wngr.

Reply:
Thanks for the input. 38290e731ae461bbeb11a0caf83c74414858088f (not yet pushed) includes the suggestions above. I am resolving here, feel free to un-resolve or start a follow up conversation.